src/base-field/base-field.tsx

```diff
@@ -137,34 +196,72 @@ type BaseFieldProps = WithEnhancedClassName &
 
 type FieldComponentProps<T extends HTMLElement> = Omit<
     BaseFieldProps,
-    'children' | 'className' | 'variant'
+    'children' | 'className' | 'fieldRef' | 'variant'
 > &
     Omit<HtmlInputProps<T>, 'className' | 'style'>
 
 function BaseField({
     variant = 'default',
     label,
+    value,
     auxiliaryLabel,
     message,
     tone = 'neutral',
     className,
     children,
     maxWidth,
+    maxLength,
     hidden,
     'aria-describedby': originalAriaDescribedBy,
     id: originalId,
 }: BaseFieldProps & BaseFieldVariantProps & WithEnhancedClassName) {
     const id = useId(originalId)
     const messageId = useId()
 
+    const inputLength = validateInputLength({ value, maxLength })
+
+    const [characterCount, setCharacterCount] = React.useState<string | null>(inputLength.count)
+    const [characterCountTone, setCharacterCountTone] = React.useState<FieldTone>(inputLength.tone)
+
     const ariaDescribedBy = originalAriaDescribedBy ?? (message ? messageId : null)
 
     const childrenProps: ChildrenRenderProps = {
         id,
+        value,
         ...(ariaDescribedBy ? { 'aria-describedby': ariaDescribedBy } : {}),
         'aria-invalid': tone === 'error' ? true : undefined,
+        onChange(event) {
+            if (!maxLength) {
+                return
+            }
+
+            const inputLength = validateInputLength({
+                value: event.currentTarget.value,
+                maxLength,
+            })
+
+            setCharacterCount(inputLength.count)
+            setCharacterCountTone(inputLength.tone)
+        },
     }
 
+    React.useEffect(
```

Comment:
I don't think an effect (which delays the update one tick) is needed for this. I think that `useMemo` should be used instead for computing the tone, which would update whenever the character count/props change.

Reply:
I believe the effect is still required to update the character count when the `value` prop changes, which happens in the following modal when clicking any of the filter example buttons:

![image](https://github.com/user-attachments/assets/f81d4f32-ff68-46ed-8794-50454e70ea1b)